0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,82 @@
+0. _(Optional)_ The receiver calls a backend to review the payment.
+0. The receiver fulfills the incoming transfer with the HMAC of the ILP packet, using the payment key as the key.
+
+## Data Format
+
+The PSK details are divided into two request objects in the style of [HTTP Requests](https://tools.ietf.org/html/rfc7230#section-3.2). The data of the outer request is the encrypted inner request.
+
+Each of the requests starts with `PSK/1.0` and then `PUBLIC` or `PRIVATE`.  The requests are made of a list of headers, followed by data. For simplicity, duplicate headers are not allowed. 
+
+The headers of the outer request are visible to all parties transmitting the request, so they SHOULD contain only the bare minimum required for the intended parties to decrypt and validate the inner request.
+
+### Outer Request
+
+The outer request is formatted as below:
+
+```http
+PSK/1.0 PUBLIC
+Key-Id: fpwpAhlN588
+Key-Algorithm: HMAC-SHA-256
+Unsafe-Header: data_everyone_can_see
+
+...
+```
+
+Everything in this request is readable by all connectors between the sender and receiver. PSK payments use an [HMAC](https://en.wikipedia.org/wiki/Hash-based_message_authentication_code) of the PSK details as a fulfillment, in order to prove that the contents are unmodified.  The pre-shared secret key is the key used in the HMAC.
+
+The data of this request is an encrypted binary blob. This encrypted blob contains the inner (PRIVATE) PSK request.
+
+Despite the fact that connectors can read the outer request's data, connectors should not be required to do so, nor should they rely on PSK details being present on all transfers. This information is end-to-end.
+
+The only headers in the outer request should be the ones required for the receiver to decrypt the inner request. In the above example, the decryption key for the outer request's body is: `HMAC-SHA-256(data: "fpwpAhlN588", key: secret)`, where `secret` is the pre-shared secret key.
+
+| Header | Value |
+|--------|-------|
+| `Key-Id` | _(Required)_ Base64url string used in the generation of a unique key for this payment |
+| `Key-Algorithm` | _(Required)_ The algorithm by which the Key-Id and pre-shared secret key generate the payment key |
+| `Unsafe-`... | _(Optional)_ Additional headers SHOULD be prefixed with `Unsafe-`. |
+
+
+### Inner Request
+
+The inner request is structured like so:
+
+```http
+PSK/1.0 PRIVATE
+Expires-At: 2017-03-13T16:58:06.901Z
+Header-A: some_stuff
+Header-B: some_other_stuff
+Header-C: some_more_stuff
+Date: 2017-03-13T16:57:56.901Z
+Content-Type: application/json
+
+{"foo":"bar"}
+```
+
+Only the intended sender and receiver can decrypt and read the inner request. The inner request contains an arbitrary message body preceded by the following headers:
+
+| Header | Value |
+|--------|-------|
+| `Expires-At` | _(Required)_ An ISO8601 UTC timestamp, after which this PSK payment has expired. |
```

Comment:
I'm not sure this should be required. If you don't set it it means that there's no expiry for the request. That's fine because the transfer you sent it on will have an expiry